### PR TITLE
set webhooks ownerReferences to namespace

### LINKED ISSUE #2044: Webhooks should get OwnerRef to namespace
/area API
/kind cleanup
/kind good-first-issue

## Expected Behavior

It's a common misconception that folks can delete our system namespace (e.g. `knative-serving`) to "clean up" their cluster for a fresh installation.

## Actual Behavior

Unfortunately, we install a lot of things that aren't deleted by deleting our system namespace, and some of those things block reinstallation.  In particular the configmap webhook has no pods, and so it blocks configmap creations.  This creates a chicken/egg problem because the webhook pods need those configmaps to start up properly.

## Proposed Solution

I think the simplest solution would be to have our webhooks set up an OwnerReference to the namespace resource within which its target service lives.  We used to do this for the webhook deployment, but removed that due to it being Cluster-scoped -> Namespaced (which isn't supported).  However, Namespace is a cluster-scoped resource, so I think this would work.

### PR DESCRIPTION
**Pull request by novahe**
# Changes

 - set webhooks ownerReferences to namespace 

/kind cleanup

Fixes #2044

**Release Note**

```release-note
NONE
```

**Docs**

```docs
```

### CODE REVIEW

**Comment by markusthoemmes** [comment]
/ok-to-test

The code looks spot on, thanks! Have we tested this is actually fixing the referenced bug? Let's also make sure the OwnerReference doesn't cause the webhook config to be garbage collected. I guess manual tests for both of those are fine.

**Comment by NovaHe** [comment]
Thank you for reviewing @markusthoemmes

In my test environment
**I tested to delete the ns, `webhook config` can be deleted:**

image

**I deleted webhook pod twice, The `webhook config` has not changed**
image
only updated once

Am I missing something? What else do I need to do?

File: webhook/configmaps/configmaps.go [hunk 1]
```diff
@@ -137,9 +137,15 @@ func (ac *reconciler) reconcileValidatingWebhook(ctx context.Context, caCert []b
 
 	webhook := configuredWebhook.DeepCopy()
 
-	// Clear out any previous (bad) OwnerReferences.
-	// See: https://github.com/knative/serving/issues/5845
-	webhook.OwnerReferences = nil
+	// Set the owner to namespace.
+	if len(webhook.OwnerReferences) == 0 {
```
Comment (markusthoemmes) [inline webhook/configmaps/configmaps.go:141]: Should we fully reconcile this? I.e. make sure it's always the correct ownerRef, even if this is non-zero (i.e. after the first reconcilation?)

File: webhook/resourcesemantics/validation/reconcile_config.go [hunk 2]
```diff
@@ -135,9 +136,15 @@ func (ac *reconciler) reconcileValidatingWebhook(ctx context.Context, caCert []b
 
 	current := configuredWebhook.DeepCopy()
 
-	// Clear out any previous (bad) OwnerReferences.
-	// See: https://github.com/knative/serving/issues/5845
-	current.OwnerReferences = nil
+	// Set the owner to namespace.
+	if len(current.OwnerReferences) == 0 {
```
Comment (markusthoemmes) [inline webhook/resourcesemantics/validation/reconcile_config.go:140]: Same q as above.

File: webhook/resourcesemantics/defaulting/defaulting.go [hunk 3]
```diff
@@ -172,9 +173,14 @@ func (ac *reconciler) reconcileMutatingWebhook(ctx context.Context, caCert []byt
 
 	current := configuredWebhook.DeepCopy()
 
-	// Clear out any previous (bad) OwnerReferences.
-	// See: https://github.com/knative/serving/issues/5845
-	current.OwnerReferences = nil
+	if len(current.OwnerReferences) == 0 {
```
Comment (markusthoemmes) [inline webhook/resourcesemantics/defaulting/defaulting.go:176]: Same q as above.

**Review by markusthoemmes — COMMENTED** [review]
/assign @dprotaso 

Sorry for having this dangling for a while. Adding Dave to assert if this isn't breaking assumptions in K8s API.

**Comment by novahe** [comment]
I fixed it, PTAL thanks

**Comment by dprotaso** [comment]
🎉 tested it out and it works

/lgtm
/approve

**Comment by julz** [comment]
This is going to avoid a _lot_ of user confusion, thanks @novahe 🎉🎉 

cc @omerbensaadon @csantanapr I think you will like!

**Pull request merged**